File: gcj_2017/QR_A.clj
; Google Code Jam 2017 Qualification Round
; Problem A. Oversized Pancake Flipper

(use '[clojure.string :only (join split includes?)])

(defn rev[s]
  (map (fn[c] (if (= c \+) \- \+)) s))

(defn f[s k t]
  (cond
    (= (count s) 0) t
    (= (first s) \+) (f (rest s) k t)
    (< (count s) k) -1
    :else (f (concat (rev (take k s)) (drop k s)) k (+ 1 t))))

(defn solve[s k]
  (def ans (f s k 0))
  (if (< ans 0)
    "IMPOSSIBLE"
    ans))

(defn output-result[t ans]
  (println (str "Case #" t ": " ans)))

(def T (read-string (read-line)))
(loop [t 1]
  (when (<= t T)
    (def x (split (read-line) #" "))
    (output-result t (solve (first x) (read-string (last x))))
    (recur (inc t))))

File: gcj_2017/QR_A.go
// Google Code Jam 2017 Qualification Round
// Problem A. Oversized Pancake Flipper

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var sc = bufio.NewScanner(os.Stdin)

func nextString() string {
	sc.Scan()
	return sc.Text()
}

func nextInt() int {
	i, e := strconv.Atoi(nextString())
	if e != nil {
		panic(e)
	}
	return i
}

func solve(s string, k int) int {
	if len(s) == 0 {
		return 0
	}
	if s[0] == '+' {
		return solve(s[1:len(s)], k)
	}
	if len(s) < k {
		return -1
	}
	a := strings.Replace(s[0:k], "+", ".", -1)
	a = strings.Replace(a, "-", "+", -1)
	a = strings.Replace(a, ".", "-", -1)
	r := solve(a + s[k:len(s)], k)
	if r < 0 {
		return -1
	} else {
		return 1 + r
	}
}

func main() {
	sc.Split(bufio.ScanWords)
	T := nextInt()
	for t := 1; t <= T; t++ {
		s := nextString()
		k := nextInt()
		ans := solve(s, k)

		fmt.Print("Case #")
		fmt.Print(t)
		fmt.Print(": ")
		if ans < 0 {
			fmt.Print("IMPOSSIBLE")
		} else {
			fmt.Print(ans)
		}
		fmt.Println("")
	}
}
